### PR TITLE
SVM: Add minibatch as a new solver

### PR DESCRIPTION
**Pull request by iyerr3**
Additional author: Nikhil Kak <nkak@pivotal.io>

This work is based on the original work by Xiaocheng Tang <xiaocheng.t@gmail.com> in #75. 

This PR adds two main features: 

1. A Minibatch solver that takes as input a batch of data
2. Minibatching for SVM

### CODE REVIEW

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/328/

File: src/modules/convex/algo/igd.hpp [hunk 1]
```diff
@@ -56,6 +59,62 @@ IGD<State, ConstState, Task>::transition(state_type &state,
...
+               X_batch = tuple.indVar.block(curr_batch_row_index, 0, batch_size, n_ind_cols);
+               y_batch = tuple.depVar.segment(curr_batch_row_index, batch_size);
+           }
+           loss += Task::getLossAndUpdateModel(
+               state.task.model, X_batch, y_batch, state.task.stepsize);
+        }
+
+        // The first epoch will most likely have the most loss.
+        // So being pessimistic, we return average loss only for the first epoch.
+        if (curr_epoch==0) state.algo.loss += loss;
```
Comment (njayaram2) [inline src/modules/convex/algo/igd.hpp]: Should we average this over `n_batches`?

File: src/ports/postgres/modules/svm/svm.py_in [hunk 2]
```diff
@@ -89,9 +113,9 @@ def _verify_table(source_table, model_table, dependent_varname,
                 "('{dependent_varname}') for source_table "
                 "({source_table})!".format(dependent_varname=dependent_varname,
                                            source_table=source_table))
-        dep_type = get_expr_type(dependent_varname, source_table)
-        if '[]' in dep_type:
-            plpy.error("SVM error: dependent_varname cannot be of array type!")
+        # dep_type = get_expr_type(dependent_varname, source_table)
```
Comment (orhankislal) [inline src/ports/postgres/modules/svm/svm.py_in]: We should remove these lines if we don't need them.

File: src/modules/convex/linear_svm_igd.cpp [hunk 3]
```diff
@@ -120,6 +124,98 @@ linear_svm_igd_transition::run(AnyType &args) {
...
+    // learning. The weight is not used for the loss computation.
+    tuple.weight = args[11].getAs<double>();
+
+
+    // Now do the transition step
+    // apply Minibatching with regularization
+    L2<GLMModel>::scaling(state.task.model, state.task.stepsize);
+    LinearSVMIGDAlgoMiniBatch::transitionInMiniBatch(state, tuple);
+    L1<GLMModel>::clipping(state.task.model, state.task.stepsize);
+
```
Comment (orhankislal) [inline src/modules/convex/linear_svm_igd.cpp]: Should we leave a comment on why the mini-batching transition step does not call the loss and gradient algorithms like the regular one?
On the other hand, I am not sure if we want to explain the lack of something in the comments. Maybe we can mention this implementation detail in the design docs?

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/330/
